src/Sylius/Bundle/CoreBundle/Validator/Constraints/CatalogPromotionScopeValidator.php

```diff
@@ -41,57 +37,20 @@ public function validate($value, Constraint $constraint): void
         Assert::isInstanceOf($constraint, CatalogPromotionScope::class);
 
         /** @var CatalogPromotionScopeInterface $value */
-        if (
-            $value->getType() !== CatalogPromotionScopeInterface::TYPE_FOR_VARIANTS &&
-            $value->getType() !== CatalogPromotionScopeInterface::TYPE_FOR_TAXONS
-        ) {
+        if (!in_array($value->getType(), $this->scopeTypes, true)) {
             $this->context->buildViolation($constraint->invalidType)->atPath('type')->addViolation();
 
             return;
         }
 
         $configuration = $value->getConfiguration();
 
-        if ($value->getType() === CatalogPromotionScopeInterface::TYPE_FOR_VARIANTS) {
-            $this->validateForVariantsType($configuration, $constraint);
-
-            return;
-        }
-
-        $this->validateForTaxonType($configuration, $constraint);
-    }
-
-    private function validateForVariantsType(array $configuration, CatalogPromotionScope $constraint): void
-    {
-        if (!array_key_exists('variants', $configuration) || empty($configuration['variants'])) {
-            $this->context->buildViolation($constraint->variantsNotEmpty)->atPath('configuration.variants')->addViolation();
-
-            return;
-        }
-
-        foreach ($configuration['variants'] as $variantCode) {
-            if (null === $this->variantRepository->findOneBy(['code' => $variantCode])) {
-                $this->context->buildViolation($constraint->invalidVariants)->atPath('configuration.variants')->addViolation();
-
-                break;
-            }
-        }
-    }
-
-    private function validateForTaxonType(array $configuration, CatalogPromotionScope $constraint): void
-    {
-        if (!isset($configuration['taxons']) || empty($configuration['taxons'])) {
-            $this->context->buildViolation($constraint->taxonsNotEmpty)->atPath('configuration.taxons')->addViolation();
-
-            return;;
-        }
-
-        foreach ($configuration['taxons'] as $taxonCode) {
-            if (null === $this->taxonRepository->findOneBy(['code' => $taxonCode])) {
-                $this->context->buildViolation($constraint->invalidTaxons)->atPath('configuration.taxons')->addViolation();
+        /** @var ScopeValidatorInterface $validator */
+        $validator = $this->scopeValidators[$value->getType()];
+        $violations = $validator->validate($configuration, $constraint);
 
-                return;
-            }
+        foreach ($violations as $path => $message) {
```

Comment:
I wonder, maybe instead of returning violations we should just use a validator and do an actual _validation_ in it? 🤔 this solution seems like a little overcomplication